### DISCUSSION TITLE
Quill Block Format Behavior Consistency

### DISCUSSION THREAD

**Discussion by aldycool** [singerdmx/flutter-quill#375] [General]
Hi @singerdmx ,

I'm currently creating a PR for this repository to add Text Alignment capability. While doing this, I've noticed a slightly weird behavior on the rendering of the Delta data. For example, it turns out that combining `align` and `header` attributes will render the `align` only. If we remove the `align`, then the `header` will be rendered correctly.

This has led me to believe, maybe some other inconsistencies can also be found somewhere in the code (I haven't check thoroughly). So, if we take a look at how originally the Quill (from QuillJS) block formats should behave, especially in their exclusivity, I've managed to summarize this information below:

***
**Quill Block Formats Behavior**
*(Checked against latest Quill JS version 1.3.6 using this [codepen.io](https://codepen.io/aldycool/pen/qBjNVGO))*

||header|indent|align|list|code-block|blockquote|direction|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|**header**|X|*Inclusive*|*Inclusive*|**Exclusive**|**Exclusive**|**Exclusive**|*Inclusive*|
|**indent**|*Inclusive*|X|*Inclusive (1)*|*Inclusive (2)*|*Inclusive*|*Inclusive*|*Inclusive (3)*|
|**align**|*Inclusive*|*Inclusive (1)*|X|*Inclusive*|*Inclusive*|*Inclusive*|*Inclusive (4)*|
|**list**|**Exclusive**|*Inclusive (2)*|*Inclusive*|X|**Exclusive**|**Exclusive**|*Inclusive*|
|**code-block**|**Exclusive**|*Inclusive*|*Inclusive*|**Exclusive**|X|**Exclusive**|*Inclusive*|
|**blockquote**|**Exclusive**|*Inclusive*|*Inclusive*|**Exclusive**|**Exclusive**|X|*Inclusive*|
|**direction**|*Inclusive*|*Inclusive (3)*|*Inclusive (4)*|*Inclusive*|*Inclusive*|*Inclusive*|X|

**Description:**

- **Exclusive:** *cannot set both attributes simulatenously*
- *Inclusive: both attributes can be set simultaneously*
- *Inclusive (1): ex.: `indent` value can offset a centered text (`align`: `center`).*
- *Inclusive (2): `indent` value determines list's numbering scheme, ex.: `indent: 1` sets the numbering scheme to 1, 2, 3..., `indent: 2` changed the numbering scheme to a, b, c..., etc.*
- *Inclusive (3): forward / backward `indent` respects the `direction`, ex.: `direction: rtl` can be indented with backward `indent`.*
- *Inclusive (4): `indent: rtl` always sets `align: right` if the `align` attribute is not present.*
- **Default:** Quill contents is in left-to-right direction, therefore the `direction: ltr` attribute is never written to Quill data.
- **Default:** Quill contents is left-aligned, therefore the `align: left` attribute is never written to Quill data. 
***

My question is: 
1. Do we (this repository) strictly follow the original exclusivity (and inclusivity) behavior above? or do we follow partially or not at all?
2. If we are not following, should we follow? 
3. If we are following, can you kindly point out which part of the code base that manage and enforce this?

Thanks!.

**Comment by singerdmx** [comment]
We should strive to follow Quill JS standard.
Our code currently may not be able to be fully compliant but Quill JS should be our goal and standard.

**Comment by singerdmx** [comment]
Code wise it is at lib/src/widgets/text_line.dart

![image](https://user-images.githubusercontent.com/122956/132079887-6060d2e2-e99b-429f-aea5-c54671a24fec.png)

**Comment by aldycool** [comment]
Hi @singerdmx, thanks for the collaborate invite. I'll fully learn the code base first. I think any modification especially about behavior and rendering should be discussed first to fully compliant with the Quill JS standard. 

This doesn't mean that we cannot add specific features that originally does not exist in Quill JS, but, at least the original behavior should be closely maintained.
